sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/BeamFnControlClient.java

```diff
@@ -102,18 +103,26 @@ public BeamFnControlClient(
     }
 
     @Override
-    public void onNext(BeamFnApi.InstructionRequest value) {
-      LOG.debug("Received InstructionRequest {}", value);
-      executor.execute(
-          () -> {
-            try {
-              BeamFnApi.InstructionResponse response = delegateOnInstructionRequestType(value);
-              sendInstructionResponse(response);
-            } catch (Error e) {
-              sendErrorResponse(e);
-              throw e;
-            }
-          });
+    public void onNext(BeamFnApi.InstructionRequest request) {
+      try {
+        BeamFnLoggingMDC.setInstructionId(request.getInstructionId());
+        LOG.debug("Received InstructionRequest {}", request);
+        executor.execute(
+            () -> {
+              try {
+                BeamFnLoggingMDC.setInstructionId(request.getInstructionId());
```

Comment:
nit: Maybe add a comment stating that the callback runs on a separate thread, which is why we need to set the instruction id here as well, might not be obvious to every reader.

Reply:
Done